AsyncGuidance.md

```diff
@@ -839,6 +840,144 @@ class AmbientValues
+
+APIs that run user callbacks usually store capture the current execution context in order to preserve async locals between callback registration and execution. Here are examples of some APIs that do this:
+
+- `Timer`
+- `CancellationToken.Register`
+- `new FileSystemWatcher`
+- `SocketAsyncEventArgs`
+- `Task.Run`
+- `ThreadPool.QueueUserWorkItem`
+
+❌ Here's an example of an execution context leak that causes memory pressure because of a lifetime mismatch between the API capturing the execution context, and the lifetime of the data stored in the async local.
+
+```C#
+using System.Collections.Concurrent;
+
+// Singleton cache
+var cache = new NumberCache(TimeSpan.FromHours(1));
+
+// Simulate 10000 concurrent requests
+Parallel.For(0, 10000, i =>
+{
+    ChunkyObject.Current = new ChunkyObject();
+
+    cache.Add(i);
+});
+
+Console.WriteLine("Before GC: " + BytesAsString(GC.GetGCMemoryInfo().HeapSizeBytes));
+Console.ReadLine();
+
+GC.Collect();
+GC.WaitForPendingFinalizers();
+
+Console.WriteLine("After GC: " + BytesAsString(GC.GetGCMemoryInfo().HeapSizeBytes));
+Console.ReadLine();
+
+static string BytesAsString(long bytes)
+{
+    string[] suffix = { "B", "KB", "MB", "GB", "TB" };
+    int i;
+    double doubleBytes = 0;
+
+    for (i = 0; bytes / 1024 > 0; i++, bytes /= 1024)
+    {
+        doubleBytes = bytes / 1024.0;
+    }
+
+    return string.Format("{0:0.00} {1}", doubleBytes, suffix[i]);
+}
+
+public class NumberCache
+{
+    private readonly ConcurrentDictionary<int, CancellationTokenSource> _cache = new ConcurrentDictionary<int, CancellationTokenSource>();
+    private TimeSpan _timeSpan;
+
+    public NumberCache(TimeSpan timeSpan)
+    {
+        _timeSpan = timeSpan;
+    }
+
+    public void Add(int key)
+    {
+        var cts = _cache.GetOrAdd(key, _ => new CancellationTokenSource());
+        // Delete entry on expiration
+        cts.Token.Register((_, _) => _cache.TryRemove(key, out _), null);
+
+        // Start count down
+        cts.CancelAfter(_timeSpan);
+    }
+}
+
+class ChunkyObject
+{
+    private static readonly AsyncLocal<ChunkyObject?> _current = new();
+
+    // Stores lots of data (but it should be gen0)
+    private readonly string _data = new string('A', 1024 * 32);
+
+    public static ChunkyObject? Current
+    {
+        get => _current.Value;
+        set => _current.Value = value;
+    }
+
+    public string Data => _data;
+}
+```
+
+The above example has a singleton `NumberCache` that stores numbers for an hour. We have a `ChunkyObject` which stores a 32K string in a field and has an async local so that any code running may access the current `ChunkyObject`. This object should be collected when the `GC` runs but instead we're implicitly capturing the `ChunkyObject` in the `NumberCache` via `CancellationToken.Register`. Instead of just caching the number and a `CancellationTokenSource`, we're implicitly capturing and storing all async locals attached to the current execution context for an hour!
```

Comment:
```suggestion
The above example has a singleton `NumberCache` that stores numbers for an hour. We have a `ChunkyObject` which stores a 32K string in a field, and has an async local so that any code running may access the current `ChunkyObject`. This object should be collected when the `GC` runs but instead, we're implicitly capturing the `ChunkyObject` in the `NumberCache` via `CancellationToken.Register`. Instead of just caching the number and a `CancellationTokenSource`, we're implicitly capturing and storing all async locals attached to the current execution context for an hour!
```